### ISSUE TITLE
Branching strategy /Git flow to align with CI/CD pipeline

### ISSUE THREAD

**Issue by s1mrankaur** [awslabs/aws-simple-cicd#13]
I would like to understand what should be the branching strategy to align well with this CI/CD pipeline. I have previously used Github flow with  Feature branch, development branch, and master branch. Since we are promoting a branch to production from the dev environment, I imagine we don't want to have dev and master as separate branches? That will also need us to approve multiple PR's for the same feature branch. 

What branching strategy works best for this CI/CD pipeline?

**Comment by gmansilla** [comment]
At this point, I think your branching startegy could be something like this:

3 active branches: master (or main), development and staging.

Master = Production code.
Development = active development. (Unstable)
Staging = code ready to be tested by QA and once approved should be merged onto master.

With this project, you would create a pipeline for each branch mentioned above.

If you're familiar with git-flow, you'll notice that with this approach, you start will all braches having the same code.

Your feature branches would be created from development and once you reach a milestone you merge all your feature branches onto Staging and have your qa validate your milestone. If approved, staging is merged onto master and released to prod. Finally, master is merged back onto staging and development. 

Another approach you can follow is to create a "release candidate" branch at the beggining of your sprints, this RC branch would be created from master and all your feature branches are created from development.

Once your feature is done, you merge onto RC and when you decide its time to test your RC, you merge RC onto staging,  if QA approves, you then merge onto master and release to prod.  After releasing, you merge (or rebase) master onto development and staging. You then delete the old RC and create a new RC from master.

With this last approach,  you would need to create a new pipeline for each new RC and destroy the old RC along with its pipeline.

It's worth mentioning that we might be adding new features to this project so that pipelines for feature branches are created automatically on the fly. It all depends on the feedback we receive (which you have already done and we thank you for that).

**Comment by srijitm** [comment]
Your branching strategy should reflect your team's culture. If the team is fast paced and releasing multiple times a week a trunk-driven approach might be best. It simplifies the branching model greatly (compared to git flow) and reduces the cognitive load for developers.

See 'Scaled Trunk-Based Development': https://trunkbaseddevelopment.com/

Out of the box this project supports Trunk-Based development where anything that is merged to master should build, pass all test and be deployed to production. 

However like @gmansilla said, depending on our customer use-cases we may look to support new branching models.

**Comment by s1mrankaur** [comment]
@srijitm Gitflow seems anti-CI/CD approach (Bundling change-sets and slow feedback). GitHub-flow seems difficult to deal with when there is a bug and the code deployed needs to be reverted.

I am not sure if my understanding is wrong but I would like to understand this in more detail with respect to this pipeline, also considering how best to manage the reverts from production and making sure the master branch is not a commit hell from merges and reverts

**Comment by s1mrankaur** [comment]
@gmansilla @srijitm  I think the suggested branching strategy does not align with the CI/CD pipeline here. 

**Talking about the first approach with three branches - Master, staging, and development:**

Here are things that confuse me:


1. If there are 5 developers in the team, Developer 1 started a pipeline for their feature branch- It gets deployed to the dev environment. Now, if developer 2 started a pipeline for another feature branch, how are deployments (2 different builds) to be tested at the same time? If they cannot be tested at the same time, then what's the benefit of using a pipeline per branch? Similarly when code from the pipeline of developer 1 gets promoted to test account, and so does from the pipeline of developer 2, what's being deployed now? How do we enable multiple developers to deploy and test their code at the same time?

2. If we are using 3 branches in this CI/CD pipeline- the code from feature_branch gets merged into development, the pipeline for the development branch gets triggered at this stage. Then it gets promoted to the "Test Account". Now because it's deployment from the development branch, we don't want to promote it to production, even if everything worked okay. Now, we merge the development branch code into the QA branch, and the pipeline for the staging branch gets triggered, goes through the development account, and test account and again because it's staging branch, we reject it and do not deploy to production. Now we go back and merge the staging branch into the master branch, and that's when the master branch pipeline gets triggered and goes all the way? Why are we repeating the same process 3 times? Did I get something wrong here?

**Comment by srijitm** [comment]
1. Modify the stack and resource names so that multiple developers can deploy their own branches into the same AWS account. This can be easily done by leveraging parameters in cloudformation, variables in terraform/cdk etc. 

2. This project is designed for trunk-based development due to its myriad benefits and low cognitive load. Like you correctly point out, complicating the build/release process doesn't add much business value. In my experience the release process is usually a reflection of the organization's culture and is a much deeper conversation.

No, parameterize the deployments so that every feature branch can deploy the same infrastructure but with different naming and tags. Every developer has access to deploy.sh in their feature branch which they can modify. Or take advantage of the environment variables made available in the CodeBuild projects.
